### DISCUSSION TITLE
How to animate barplot and mapping in Shiny?

### DISCUSSION THREAD

**Discussion by Saupia** [teamalgoritma/community#436] [Q&A] [answered]
Masalah:
1. Animasi yang sudah saya buat tidak bisa beroperasi di Shiny.
2. Menu Map tidak bisa dibuka. 

**Server**

`shinyServer(function(input, output) { 
  
  output$plot_1 <- renderPlotly({
    region <- 
      indolang %>% 
      group_by(Region) %>% 
      summarise(sum_Tot_Lang = sum(Tot_lang))
    # Pembuatan visual statis
    plot1 <- 
      ggplot(region,aes(x = Region,
                        y = sum_Tot_Lang,
                        fill = sum_Tot_Lang)) +
      geom_col() + 
      scale_fill_gradient(low = "black", high = "darkred") +
      theme_minimal() +
      labs(title = "Total Languages Based on the Regions",
           x = NULL, y = NULL) +
      theme(plot.title = element_text(face = "bold", size = 18)) +
      theme(legend.position = "none")
    plot1 + transition_states(Region, wrap = FALSE) +
      shadow_mark() + 
      enter_grow() +
      enter_fade()
  })
  #input$select_category digunakan untuk mengubah semua menu categori secara interaktif.
  #ketika di server, koma tidak harus ditambahkan
  output$plot_2 <- renderPlotly({
    egids_tot <- indolang %>% 
      filter(Region %in% input$select_category) %>% 
      group_by(EGIDS_Label) %>% 
      summarise(sum_Tot_lang = sum(Tot_lang))
      # mutate(label = glue("EGIDS Scale: {EGIDS_Label}
      #                 Total Languages: {scales::comma(sum_Tot_lang)}"))
    # Visual statis
    plot2 <- 
      egids_tot %>% 
      arrange(sum_Tot_lang) %>% 
      mutate(EGIDS_Label = factor(EGIDS_Label, levels = EGIDS_Label)) %>% 
      ggplot(aes(x = EGIDS_Label, y = sum_Tot_lang)) +
      geom_segment(aes(xend = EGIDS_Label, yend = 0))+
      geom_point(size = 4, color = "red") +
      coord_flip() +
      labs(title = "The language vitality in Indonesia", x = NULL, y = "Number of languages")
    
    #visual interactive
    # Please type your code here
    ggplotly(p = plot2, tooltip = "text")
    
  })
  
  output$plot_3 <- renderPlotly({
    devs <- indolang %>% 
      filter(Region %in% input$select_category) %>% 
      group_by(Dev_scale) %>% 
      summarise(sum_Tot_lang = sum(Tot_lang)) %>% 
      mutate(label = glue("Development: {Dev_scale}
                      Total language: {scales::comma (sum_Tot_lang)}"))
    #
    plot3 <- 
      ggplot(devs, aes(x = Dev_scale,
                                y = sum_Tot_lang,
                                fill = Dev_scale)) +
      geom_bar(stat = "identity") +
      theme(legend.position = "none") +
      labs(x = NULL, y = NULL) +
      theme(plot.title = element_text(face = "bold", size = 18))
    plot3 + transition_states(Dev_scale, wrap = TRUE) +
      shadow_mark() + 
      enter_grow() +
      enter_fade()
    #
    ggplotly(p = plot3, tooltip = "text")
    
  })
  
  output$map <- renderLeaflet({
    egids7 <-indolang %>% 
      filter(indolang$EGIDS %in% "7")
    egids8a <- indolang %>% 
      filter(indolang$EGIDS %in% "8a")
    egids8b <- indolang %>% 
      filter(indolang$EGIDS %in% "8b")
    egids9 <- indolang%>% 
      filter(indolang$EGIDS %in% "9")
    
    map <- leaflet() %>% 
      setView(lng=117, lat= -0.7, zoom=4, options = "permanent" ) %>% 
      addProviderTiles("Esri.WorldImagery", group = "Esri Imagery") %>% 
      addTiles(options = providerTileOptions(noWrap = TRUE), group="Open Street") %>% 
      addCircleMarkers(data = egids7, lng=~Longitude , lat=~Latitude, radius=8 , color="black",
                       fillColor="yellow", stroke = TRUE, fillOpacity = 0.8, group="Shifting", popup = ~Language_Name ) %>% 
      addCircleMarkers(data = egids8, lng=~Longitude , lat=~Latitude, radius=8 , color="black",
                       fillColor="red", stroke = TRUE, fillOpacity = 0.8, group="Moribund", popup = ~Language_Name) %>% 
      addCircleMarkers(data = egids8b,lng=~Longitude , lat=~Latitude, radius=8 , color="black",
                       fillColor="grey", stroke = TRUE, fillOpacity = 0.8, group="Nearly Extinct", popup = ~Language_Name) %>% 
      addCircleMarkers(data = egids9,lng=~Longitude , lat=~Latitude, radius=8 , color="black",
                       fillColor="black", stroke = TRUE, fillOpacity = 0.8, group="Dormant", popup = ~Language_Name) %>% 
      addLayersControl(overlayGroups = c("Shifting","Moribund", "Nearly Extinct", "Dormant") , baseGroups = c("Esri Imagery", "Open Street"), 
                       options = layersControlOptions(collapsed = FALSE)) 
              options = list(scrollX = TRUE)
  
    
  })
  
  

  })`

**UI**

`# Fungsi dashboardPage() diperuntuhkan untuk membuat ketiga bagian pada Shiny
dashboardPage(skin = "red",
  
  # Fungsi dashboardHeader() adalah bagian untuk membuat header
  dashboardHeader(title = "Indonesia"),
  
  # Fungsi dashboardSidebar() adalah bagian untuk membuat sidebar
  dashboardSidebar(
    
    sidebarMenu(
      menuItem(
        text = "Introduction", #Animate the total language regionally
        tabName = "introduction",
        icon = icon("magnifying-glass-chart")
      ),
      
      menuItem(
        text = "Scales",
        tabName = "scales",
        icon = icon("scale-balanced")
      ),
      
      menuItem(
        text = "Maps",
        tabName = "map",
        icon = icon("location-pin-lock")
      )
    )
    
  ),
  
  # Fungsi dashboardBody() adalah bagian untuk membuat isi body
  dashboardBody(
    tabItems(
      
      tabItem(tabName = "introduction",
              
              fluidRow(
                shinydashboard::valueBox("Numbers of languages",
                        value = length(unique(indolang$ISO)),
                        icon = icon("info"),
                        color = "light-blue"),
                
                shinydashboard::valueBox("Total language development",
                        value = length(unique(indolang$Literature)),
                        icon = icon("book"),
                        color = "light-blue"),
                
                shinydashboard::valueBox("Total of material produce ",
                        value = sum(indolang$Tot_lit),
                        icon = icon("newspaper"),
                        color = "light-blue"),
              ),
              
              fluidRow(
                
                plotlyOutput(outputId = "plot_1")
              )
              
              ),
      tabItem(tabName = "scales",
              fluidRow(
                #box menjadi wadah untuk mengatur lebar dan tinggi
                box(width = 12,
                    selectInput(inputId = "select_category", 
                                label = "Choose the Region (please click the box below)", 
                                choices = unique(indolang$Region))
                    
                    )
              ),
              fluidRow(
                #maksimal lebar kotak adalah 12. Untuk membagi dua maka total lebar dibagi 2 sama dengan 6. 
                box(width = 6,
                    plotlyOutput(outputId = "plot_2")),
                
                box(width = 6,
                    plotlyOutput(outputId = "plot_3"))  
              )
              
              ),
      tabItem(fluidPage(leafletOutput("map"),
                        p(),
                        actionButton("recalc", "New points")
              )
    )
  )
))`

**GLOBAL**

`# Untuk menyiapkan lingkungan aplikasi, misalnya, library, impor data dan persiapan data

# --------- LOAD LIBRARIES

# dashboarding
library(shiny)
library(shinydashboard)
library(DT) # datatable

options(scipen = 99) # me-non-aktifkan scientific notation
library(tidyverse) # koleksi beberapa package R
library(plotly) # plot interaktif
library(glue) # setting tooltip
library(scales) # mengatur skala pada plot
library(ggplot2) # for plotting
library(gganimate) # visual animation
library(leaflet)
library(gapminder)
library(gifski)
library(ggplot2)
library(geojsonio)
library(glue)
library(shinyanimate)

# --------- READ DATA 

vids <- indolang <- read.csv("indolang.csv")

# --------- DATA PREPARATION

indolang$Speakers <- as.numeric(gsub(",","",indolang$Speakers))
`

**Comment by ramdhinafinita** [comment, marked as answer]
Dear Bapak @Saupia,

Berdasarkan pertanyaan yang Bapak tanyakan yakni:

1. Animasi yang sudah saya buat tidak bisa beroperasi di Shiny.
Jawaban: disetiap ggplot aes() silakan ditambah paramater text=label
```
plot1 <-
ggplot(region,aes(x = Region,
y = sum_Tot_Lang,
fill = sum_Tot_Lang,
text=label))
```
3. Menu Map tidak bisa dibuka.
Jawaban: Pada file Server.R pada map leaflet terdapat salah koding **data = egids8** seharusnya **data = egids8a**
```
addCircleMarkers(data = egids8a, lng=~Longitude , lat=~Latitude, radius=8 , color="black",
                   fillColor="red", stroke = TRUE, fillOpacity = 0.8, group="Moribund", popup = ~Language_Name) %>% 
```

Jika jawaban ini dirasa membantu, silahkan tandai sebagai "Mark As Answer". Terimakasih
